### PR TITLE
DS-4124: Add ITs to prove access restricted Bitstreams cannot be accessed anonymously

### PR DESCRIPTION
**Pull request by Micheleboychuk**
See https://jira.lyrasis.org/browse/DS-4124
Please note that this PR is build on top of https://github.com/DSpace/DSpace/pull/2687

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 1]
```diff
@@ -349,6 +351,170 @@ public void testEmbargoedBitstream() throws Exception {
...
+
+            Bitstream bitstream = BitstreamBuilder
+                .createBitstream(context, publicItem1, is)
+                .withName("Test Embargoed Bitstream")
+                .withDescription("This bitstream is embargoed")
+                .withMimeType("text/plain")
+                .withEmbargoPeriod("3 months")
+                .build();
+
+            context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Can we move this context.restoreAuthSystemState() and everything after this *outside* of the `try` clause?  I think the `try` only needs to be around the `BitstreamBuilder`, unless I'm misunderstanding something?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 2]
```diff
@@ -349,6 +351,170 @@ public void testEmbargoedBitstream() throws Exception {
...
+
+            Bitstream bitstream = BitstreamBuilder
+                .createBitstream(context, publicItem1, is)
+                .withName("Test Embargoed Bitstream")
+                .withDescription("This bitstream is embargoed")
+                .withMimeType("text/plain")
+                .withEmbargoPeriod("-3 months")
+                .build();
+
+            context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Again, here, I think the `try` can be closed *after* the BitstreamBuilder. Everything else can move outside of the `try` clause. Look at how you did this in `embargoedBitstreamAccessGrantByAdminsTest()`

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 3]
```diff
@@ -349,6 +351,170 @@ public void testEmbargoedBitstream() throws Exception {
...
+                .withName("Test Embargoed Bitstream")
+                .withDescription("This bitstream is embargoed")
+                .withMimeType("text/plain")
+                .withEmbargoPeriod("3 months")
+                .build();
+
+            context.restoreAuthSystemState();
+
+            String authToken = getAuthToken(eperson.getEmail(), password);
+            getClient(authToken).perform(get("/api/core/bitstreams/" + bitstream.getID() + "/content"))
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Please add a similar test for an unauthenticated user.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 4]
```diff
@@ -349,6 +351,170 @@ public void testEmbargoedBitstream() throws Exception {
...
+        String tokenAdminCollection1 = getAuthToken(adminCollection1.getEmail(), "qwerty03");
+        getClient(tokenAdminCollection1).perform(get("/api/core/bitstreams/" + bitstream.getID() + "/content"))
+                   .andExpect(status().isOk());
+
+        checkNumberOfStatsRecords(bitstream, 2);
+
+        // admin of second collection is NOT allowed access to embargoed item  of first collection
+        String tokenAdminCollection2 = getAuthToken(adminCollection2.getEmail(), "qwerty01");
+        getClient(tokenAdminCollection2).perform(get("/api/core/bitstreams/" + bitstream.getID() + "/content"))
+                   .andExpect(status().isForbidden());
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Please add a similar Forbidden test for a Community Admin of a (new) Community which is not a parent Community of collection1.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 5]
```diff
@@ -396,6 +562,141 @@ public void testPrivateBitstream() throws Exception {
...
+                    .withIssueDate("2013-01-17")
+                    .withAuthor("Doe, John")
+                    .build();
+            Bitstream bitstream = BitstreamBuilder
+                .createBitstream(context, item, is)
+                .withName("Test Embargoed Bitstream")
+                .withDescription("This bitstream is embargoed")
+                .withMimeType("text/plain")
+                .withReaderGroup(restrictedGroup)
+                .build();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Again, I think you can close the `try` clause here.  Also, you are missing a `context.restoreAuthSystemState();` in this test.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 6]
```diff
@@ -396,6 +562,141 @@ public void testPrivateBitstream() throws Exception {
...
+                    .withIssueDate("2018-10-17")
+                    .withAuthor("Doe, John")
+                    .build();
+            Bitstream bitstream = BitstreamBuilder
+                .createBitstream(context, item, is)
+                .withName("Test Embargoed Bitstream")
+                .withDescription("This bitstream is embargoed")
+                .withMimeType("text/plain")
+                .withReaderGroup(restrictedGroup)
+                .build();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: Again, I think you can close the try clause here. Also, you are missing a context.restoreAuthSystemState(); in this test too.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 7]
```diff
@@ -396,6 +562,141 @@ public void testPrivateBitstream() throws Exception {
...
+            String tokenAdminCol1 = getAuthToken(adminCol1.getEmail(), "qwerty01");
+            getClient(tokenAdminCol1).perform(get("/api/core/bitstreams/" + bitstream.getID() + "/content"))
+                                .andExpect(status().isOk());
+
+            checkNumberOfStatsRecords(bitstream, 2);
+
+            // collection2's admin user is NOT allowed access to the item belong collection1
+            String tokenAdminCol2 = getAuthToken(adminCol2.getEmail(), "qwerty02");
+            getClient(tokenAdminCol2).perform(get("/api/core/bitstreams/" + bitstream.getID() + "/content"))
+                                .andExpect(status().isForbidden());
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: It'd be good to again add a second Community that is not a parent of collection1, and ensure that Community's Admin is also Forbidden.

**Review by tdonohue — COMMENTED** [review]
@Micheleboychuk :  I reviewed this as well. The tests look good overall, but I have some suggestions for minor improvements & additional tests.

**Comment by abollini** [comment]
@tdonohue  this should be ready to go now

**Review by tdonohue — APPROVED** [review]
👍 Looks good now, thanks @Micheleboychuk and @abollini .  Again, this is just ITs (built on top of #2687 which was just merged).  This also can be merged immediately as it was flagged for a single review by the DSpace 7 WG

**Pull request merged**